src/v/archival/async_data_uploader.h

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cluster/partition.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "utils/retry_chain_node.h"
+#include "vassert.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/rwlock.hh>
+#include <seastar/core/scheduling.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <boost/iterator/counting_iterator.hpp>
+#include <boost/iterator/transform_iterator.hpp>
+#include <boost/outcome.hpp>
+#include <boost/outcome/outcome.hpp>
+#include <boost/outcome/policy/all_narrow.hpp>
+
+#include <exception>
+#include <utility>
+
+namespace archival {
+
+namespace detail {
+inline model::offset cast_to_offset(int64_t i) noexcept {
+    return model::offset(i);
+}
+} // namespace detail
+
+/// Representation of the inclusive monotonic offset range
+struct inclusive_offset_range {
```

Comment:
I'd prefer having an `offset_range` type, and if you want to avoid having to do the `x + 1` math everywhere, add a static method that builds an inclusive range so you can call it like `offset_range::inclusive(a, b)`.

One reason is consistency with all kind of ranges in C++.
Second reason is that `size()` will begin to make sense for default constructed ranges. Currently, `inclusive_offset_range{}.size() == 1` which doesn't make sense.
Third one,